### PR TITLE
🏃 Add e2e tests for all flavors

### LINKED ISSUE #577: E2E test using Cluster API test framework
/kind feature

**Describe the solution you'd like**

We can update E2E test to use Cluster API test framework.

**Anything else you would like to add:**

Cluster API test framework is guided in [The Cluster API Book](https://cluster-api.sigs.k8s.io/developer/e2e.html).

### LINKED ISSUE #723: OpenStack 'port' does not get deleted if VM failed to create
/kind bug

**What steps did you take and what happened:**
* I created a cluster but had forgotten to create the openstack `keypair` beforehand.
* This led to VM not being created with error: `Invalid key_name provided` from openstack.
* Thus, deleted the cluster with `kubectl delete cluster <my-cluster>`.
* This resulted in errors in the controller-manager pod, complaining that the security group could not be deleted because it is in use.
* The reason for `security group in use` was that the `port` had been created for the VM. But since the VM was never created, the `port` was never attached and thus never got deleted.

**What did you expect to happen:**
If VM is not created for any reason (maybe invalid image name provided, invalid flavor provided, invalid keypair name provided etc.) then the openstack `port` which was created for VM but was never used, should be deleted when trying to delete the cluster.

**Anything else you would like to add:**
Ref: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/controllers/openstackmachine_controller.go#L221-L226
There should be a check here for existing ports too at the time of VM deletion. Maybe a `InstancePortExists`, just like the `InstanceExists` function?

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built):
- Cluster-API version:
- OpenStack version:
- Minikube/KIND version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #741: Can not delete cluster due to SecurityGroupInUse
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

1. Create a workload cluster
2. Delete the workload cluster

**What did you expect to happen:**

Deleting the cluster is not completed.
OpenStack instances is deleted.
But the following resources are not deleted.

 * network
 * security groups

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

A cause of the trouble is failure of security group deletion.
See the following logs:

```
$ k -n capo-system logs capo-controller-manager-7595db778f-9rl8n manager

--- snip ---

E0215 03:34:29.362070       1 controller.go:257] controller-runtime/controller "msg"="Reconciler error" "error"="failed to delete security group: Expected HTTP response code [] when accessing [DELETE https://172.16.71.80:9696/v2.0/security-groups/b9d2ad4d-dfc3-4f27-8a33-7bbba21b3cb3], but got 409 instead\n{\"NeutronError\": {\"message\": \"Security Group b9d2ad4d-dfc3-4f27-8a33-7bbba21b3cb3 in use.\", \"type\": \"SecurityGroupInUse\", \"detail\": \"\"}}" "controller"="openstackcluster" "name"="capi-quickstart" "namespace"="default"
```

Deleting the security group for worker nodes was failed.
The security group was applied to ports.
So the ports must be deleted before deleting the security groups.

Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): `v0.3.3`
- Cluster-API version: `v0.3.13`
- OpenStack version: `Queens`
- Minikube/KIND version: `kind v0.9.0 go1.15.2 linux/amd64`
- Kubernetes version (use `kubectl version`): `v1.18.15`
- OS (e.g. from `/etc/os-release`): 

```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
This PR adds e2e tests for all our flavors (and some additional ones).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #577
Fixes #723
Fixes #741

Notes:

This PR can also be reviewed but won't be merged before the test framework and the linter PR and also contains those commits/PRs

- [x] squashed commits
- if necessary:
  - [x] includes documentation
  - [x] adds unit tests

/hold

### CODE REVIEW

**Comment by sbueringer** [comment]
@jichenjc @hidekazuna @prankul88 Pull request can be reviewed now, please take a look :)

**Comment by jichenjc** [comment]
several commits exist, should we merge from first PR?

File: hack/ci/devstack-cloud-init.yaml.tpl
Comment (jichenjc) [inline hack/ci/devstack-cloud-init.yaml.tpl:83]: maybe a TODO here

**Comment by sbueringer** [comment]
> several commits exist, should we merge from first PR?
> 
> do you have a way to extract the changes _only_ for this PR?

In this order:
* https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/782
* https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/797
* https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/798

I'll drop the duplicate commits here after #782 and #797 are merged. Just wanted to build on top and test the combination of this stuff.

For review there's a selection box in the files tab where you can select the commits of only this PR:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/798/files/97ff55feb541d83c9a33b808c49601cb6a58859c..55dd934763a30193525f27d33ec5c8912bc492b8

File: hack/ci/devstack-cloud-init.yaml.tpl
Comment (sbueringer) [reply hack/ci/devstack-cloud-init.yaml.tpl:83]: I would keep it for now. This adds a dependency but we already have the dependency because of the ubuntu image but this issue will try to fix it: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/794

(it just takes a long time to build the amphora image during the devstack installation on each test)

**Comment by sbueringer** [comment]
/test pull-cluster-api-provider-openstack-make-conformance
/test pull-cluster-api-provider-openstack-e2e-test

Also dropped the commit here, so ready for review/merge now.

@jichenjc @hidekazuna @prankul88 PTAL, if you have some time :)

File: controllers/openstackcluster_controller.go
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:188]: This should fix: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/723

File: controllers/openstackcluster_controller.go
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:184]: This should fix: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/723

**Comment by sbueringer** [comment]
Some final polishing (and subsequent linter fixes...), but now I'm done.
/test pull-cluster-api-provider-openstack-e2e-test
/test pull-cluster-api-provider-openstack-conformance-test

File: pkg/cloud/services/networking/network.go
Comment (jichenjc) [inline pkg/cloud/services/networking/network.go:389]: I'd think we at least log this behavior

File: pkg/cloud/services/networking/service.go
Comment (jichenjc) [inline pkg/cloud/services/networking/service.go:52]: is it possible clientOpts.AuthInfo is nil? 
if so the projectID will be "" , will it impact follow up features?

File: test/e2e/shared/exec_test.go
Comment (jichenjc) [inline test/e2e/shared/exec_test.go:51]: I am not sure why we need such change? do we have hard dependency on IP?

**Review by jichenjc — COMMENTED** [review]
overall quick , some nit question

File: pkg/cloud/services/networking/network.go
Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:389]: Yup I"ll add log statements where we do (i.e. delete) something

File: pkg/cloud/services/networking/service.go
Comment (sbueringer) [reply pkg/cloud/services/networking/service.go:52]: I think it cannot be nil, but it has very high impact if it is. With an empty project id we would just list and delete orphaned ports depending on the rights of the user.

This means when you have some kind of admin / domain admin credentials in the cloud.yaml it will delete stuff all over the OpenStack.

I think we should fail hard here if we're unable to determine a project id

File: test/e2e/shared/exec_test.go
Comment (sbueringer) [reply test/e2e/shared/exec_test.go:51]: This was just for debugging locally. The test is not executed (t.Skip() above)
It's just way easier to develop the exec method locally when you don't have to  execute the whole e2e tests to do it :)

I'll rollback the IPs, I just changed them when debugging the func

File: hack/ci/devstack-on-gce-project-install.sh
Comment (sbueringer) [inline hack/ci/devstack-on-gce-project-install.sh:59]: Ah I got this from testing locally. I would like to keep it anyway. The idea is to not automatically create a lot of subnets as GCP is doing it with subnet mode auto

File: test/e2e/shared/exec_test.go
Comment (jichenjc) [reply test/e2e/shared/exec_test.go]: I am just curious as, not a big deal

File: pkg/cloud/services/networking/service.go
Comment (jichenjc) [reply pkg/cloud/services/networking/service.go]: ok, thanks for the info

**Comment by sbueringer** [comment]
/test pull-cluster-api-provider-openstack-e2e-test
/test pull-cluster-api-provider-openstack-conformance-test

@jichenjc thx for the review, good findings :) Should be all fixed now ptal

File: test/e2e/shared/exec_test.go
Comment (sbueringer) [reply test/e2e/shared/exec_test.go]: Absolutely fine :)

**Comment by sbueringer** [comment]
/test pull-cluster-api-provider-openstack-e2e-test
/test pull-cluster-api-provider-openstack-conformance-test

(small fixup, was necessary because of the change to the subnet)

**Comment by jichenjc** [comment]
@sbueringer please help to squash commit then I think we might merge this

**Comment by sbueringer** [comment]
> @sbueringer please help to squash commit then I think we might merge this

@jichenjc done :)

File: pkg/cloud/services/networking/network.go
Comment (rustycl0ck) [inline pkg/cloud/services/networking/network.go:386]: We shouldn't delete anything which is not created by ClusterAPI. This statement would lead to deletion of any ports which were not created by ClusterAPI but were created for other purposes (either manually or via some other automation).

One common use case for this: In cases where a loadbalancer is not available (or the user does not use if for whatever reasons), and instead a VIP (Virtual IP) along with `keepalived` is used for load balancing, then an openstack port is created (to reserve the IP and to create proper routes), but is not attached to any VM. Since this openstack port is not attached to any VM, it has `device_owner == ""`.

The `DeleteOrphanedPorts` function would delete this port unnecessarily and would cause service disruption elsewhere.

**Review by rustycl0ck — CHANGES_REQUESTED** [review]

File: pkg/cloud/services/networking/network.go
Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:386]: @rustycl0ck Understood. So I assume there is no real way to cleanup those ports safely. I'll drop this for now from the e2e tests. Not sure how we can get rid of the problems during cluster deletion though. I'll try if I can get the e2e tests clean without this change but I'm not sure if it's possible

Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:386]: I'm not entirely sure about the general statement "we shouldn't delete anything which is not created by ClusterAPI". In my opinion there are cases where it's totally valid. One case is that the cluster as a whole should be deleted and somebody has the clean up the loadbalancers which has been created by the cloud provider openstack. This is also done in CAPA and the ccm itself is unable to delete them as it doesn't exist anymore.

Are those service interruptions really an issue when the whole cluster is deleted?

Comment (rustycl0ck) [reply pkg/cloud/services/networking/network.go:386]: My use case was referring to a setup that I worked on, where a single openstack network (IPv6) was being used by multiple teams/products. While I was trying to use CAPI to setup a k8s cluster there, a few other teams had used the VIP approach for making their applications HA (since a loadbalancer wasn't available). Just FYI, the other products weren't k8s based, but legacy VM based systems/appliccations.

#723 was caused at the time of cluster creation where the port is created first, but if the image (or the ssh key or the flavor) does not exist, server creation would error out and the port would be left behind without any cleanup on server creation failure. This issue seems to have been addressed in #705 correctly (where a rollback is performed if the server creation failed). Similarly, we should try to fix the root cause instead of performing a general cleanup.

I think the correct way to fix this would be to figure out why this error occurred in the first place.
From the following code block: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/services/compute/instance.go#L631-L648, I don't see how/why the port would have been left behind while the server got deleted successfully. If we can find the root cause, we can fix the problem in a cleaner way.

Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:386]: @rustycl0ck Thx providing some more details :). 

I think overall you're right. The better approach would be to cleanup the ports wherever they orphaned. I'm currently testing if this PR also works without the `DeleteOrphanedPorts` func: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/816
I assume this will be successful and then I'll drop the func on the current PR.

Out of curiosity (I'm not sure what is possible with OpenStack). Did you have:
* a separate project and where using the same project and network as the other teams or
* have you been using the same network but separate projects?

Comment (rustycl0ck) [reply pkg/cloud/services/networking/network.go:386]: It was a single project and same network, because all the applications (k8s hosted as well as legacy VM based) had to interact with each other (on a private network internally).

Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:386]: As I feared the tests are now failing because they are timing out because the cluster deletion does not work: https://prow.k8s.io/log?container=test&id=1376784008490258432&job=pull-cluster-api-provider-openstack-e2e-test

I would suggest merging the PR anyway without the DeleteOrphanedPort func. I would rather have those tests with a known issue on master and unblock a lot of other PRs instead of having to rely only on the conformance tests.Those tests are not mandatory yet and I try to find the bug in a follow-up PR.

What do you think?
(@jichenjc )

Comment (sbueringer) [reply pkg/cloud/services/networking/network.go:386]: Got the exact results. Looks like the error only happens when server are not successfully created right now. I'll check if I can just fix this...

Comment (sbueringer) [reply pkg/cloud/services/networking/network.go]: Fixed it here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/798/commits/2e012c6e8fbbf51a844f9f514c62544c5ff998b1
Local tests were green. I'll squash and re-run the tests on this PR

**Comment by jichenjc** [comment]
/lgtm 

thanks :)

**Comment by sbueringer** [comment]
tests are green, so:
/hold cancel

**Pull request merged**
